### PR TITLE
pybind/mgr/progress: Disregard unreported pgs

### PR DESCRIPTION
**Pull request by kamoltat**
The global recovery event progress calculations only
takes into account pgs with `reported_epoch < start_epoch_of_event`
but sometimes the pgs doesn't get move before or after the creation
of the global recovery event, therefore this might result in a bug
where the global event gets stuck forever unless there is another
event that specifically makes the pgs that get stuck moves and updates
its `reported_epoch`.

Therefore, we decided to disregard pgs that are in active+clean state
but has `reported_epoch < start_epoch_of_event`.

Fixes: https://tracker.ceph.com/issues/49988

Signed-off-by: Kamoltat <ksirivad@redhat.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/pybind/mgr/progress/module.py [hunk 1]
```diff
@@ -194,15 +194,17 @@ def __init__(self, message, refs, add_to_ceph_s, start_epoch, active_clean_num):
...
         # type: (Dict) -> None
         "Update progress of Global Recovery Event"
 
+        skipped_pgs = 0
         pgs = pg_dump['pg_stats']
         new_active_clean_num = 0
         for pg in pgs:
 
             if int(pg['reported_epoch']) < int(self._start_epoch):
+                skipped_pgs += 1
```
Comment (jdurgin) [inline src/pybind/mgr/progress/module.py:207]: was this caused by the test deleting pools, and those old stats sticking around too long?

If so, it'd be better to test for the pool still existing here, so we don't complete an event that hasn't had pg updates from non-active+clean pgs yet, but is still in progress. Better yet, we could delete removed pools from pg stats as soon as the pool is deleted - the OSD no longer sends updates after a pool is deleted

**Review by jdurgin — CHANGES_REQUESTED** [review]
Glad you were able to figure out the failure! I think we should fix it within the mgr stats handling itself, rather than the modules

File: src/pybind/mgr/progress/module.py (hunk 1 above)
Comment (jdurgin) [reply src/pybind/mgr/progress/module.py:207]: looking at the mgr handling, it should be immediately removing pg stats when it handles the osdmap deleting the pool. perhaps there's a delay before the progress module notices because it's only updating the global progress event when there's a new pg_summary available, not when there's a new osdmap. rather than add more processing to a hot path, perhaps it's time to switch the progress module to update periodically instead of every time there's a new osd or pg map.

**Comment by tchaikov** [comment]
jenkins test make check

**Comment by tchaikov** [comment]
jenkins test api

**Comment by tchaikov** [comment]
jenkins test api

**Comment by kamoltat** [comment]
jenkins test api

**Comment by kamoltat** [comment]
jenkins test make

**Comment by kamoltat** [comment]
jenkins run make check

**Comment by kamoltat** [comment]
jenkins do make check

**Comment by kamoltat** [comment]
https://pulpito.ceph.com/ksirivad-2021-05-18_19:46:54-rados:mgr-wip-ksirivad-fix-bug-49988-distro-basic-smithi/

File: src/pybind/mgr/progress/module.py [hunk 2]
```diff
@@ -23,6 +23,7 @@
 
 
 class Event(object):
+
```
Comment (tchaikov) [inline src/pybind/mgr/progress/module.py:26]: why?

File: src/pybind/mgr/progress/module.py [hunk 3]
```diff
@@ -194,21 +195,31 @@ def __init__(self, message, refs, add_to_ceph_s, start_epoch, active_clean_num):
         self._active_clean_num = active_clean_num
         self._refresh()
 
-    def global_event_update_progress(self, pg_dump):
-        # type: (Dict) -> None
+    def global_event_update_progress(self, pg_dump, log):
+        # type: (Dict, Any) -> None
```
Comment (tchaikov) [inline src/pybind/mgr/progress/module.py:199]: s/Any/logging.Logger/

File: src/pybind/mgr/progress/module.py [hunk 4]
```diff
@@ -194,21 +195,31 @@ def __init__(self, message, refs, add_to_ceph_s, start_epoch, active_clean_num):
...
+            # Disregard PGs that are not being reported
+            # if the states are active+clean. Since it is
+            # possible that some pgs might not have any movement
+            # even before the start of the event.
 
             state = pg['state']
 
             states = state.split("+")
 
+            if int(pg['reported_epoch']) < int(self._start_epoch):
```
Comment (tchaikov) [inline src/pybind/mgr/progress/module.py:216]: why do we need to cast these variables to `int`?

File: src/pybind/mgr/progress/module.py [hunk 5]
```diff
@@ -301,7 +313,8 @@ def pg_update(self, raw_pg_stats, pg_ready, log):
         # FIXME: far more fields getting pythonized than we really care about
         # Sanity check to see if there are any missing PGs and to assign
         # empty array and dictionary if there hasn't been any recovery
-        pg_to_state = dict([(p['pgid'], p) for p in raw_pg_stats['pg_stats']]) # type: Dict[str, Any]
+
```
Comment (tchaikov) [inline src/pybind/mgr/progress/module.py:316]: i'd avoid adding empty lines unless it helps with the readability.

File: src/pybind/mgr/progress/module.py [hunk 6]
```diff
@@ -301,7 +313,8 @@ def pg_update(self, raw_pg_stats, pg_ready, log):
...
+        pg_to_state = dict([(p['pgid'], p) for p in raw_pg_stats['pg_stats']])  # type: Dict[str, Any]
```
Comment (tchaikov) [inline src/pybind/mgr/progress/module.py:317]: no need to materialize the generator to a list before converting them to a dict. you can just put
```py
pg_to_state = dict((p['pgid'], p) for p in raw_pg_stats['pg_stats'])
```

File: src/pybind/mgr/progress/module.py [hunk 7]
```diff
@@ -434,10 +447,10 @@ class Module(MgrModule):
             runtime=True
         ),
         Option(
-            'persist_interval',
+            'sleep_interval',
             default=5,
             type='secs',
-            desc='how frequently to persist completed events',
+            desc='how frequently to wake up and update events',
```
Comment (tchaikov) [inline src/pybind/mgr/progress/module.py:453]: nit, interval is not identical to frequency. the larger the former the smaller the latter. so the combination of the document and the name of variable is confusing.

File: src/pybind/mgr/progress/module.py [hunk 8]
```diff
@@ -451,7 +464,7 @@ def __init__(self, *args, **kwargs):
         super(Module, self).__init__(*args, **kwargs)
 
         self._events = {}  # type: Dict[str, Union[RemoteEvent, PgRecoveryEvent, GlobalRecoveryEvent]]
-        self._completed_events = [] # type: List[GhostEvent]
+        self._completed_events = []  # type: List[GhostEvent]
```
Comment (tchaikov) [inline src/pybind/mgr/progress/module.py:467]: this is not part of this change.

File: src/pybind/mgr/progress/module.py [hunk 9]
```diff
@@ -607,53 +620,59 @@ def _pg_state_changed(self, pg_dump):
         except ZeroDivisionError:
             return
         if progress < 1.0:
+            self.log.warning("Starting Global Recovery Event,\
+              {0} pgs not in active + clean state \
+              ".format(total_pg_num - active_clean_num))
```
Comment (tchaikov) [inline src/pybind/mgr/progress/module.py:625]: nit, i'd suggest to be more consistent when printing logging message, we have
```py
log.debug("Updated progress to %s", self.summary())
```
couple lines above. so, would be better to put
```py
self.log.warning(("Starting Global Recovery Event,"
                  "%d pgs not in active + clean state"),
                 total_pg_num - active_clean_num)
```

File: src/pybind/mgr/progress/module.py [hunk 10]
```diff
@@ -607,53 +620,59 @@ def _pg_state_changed(self, pg_dump):
...
 
-    def notify(self, notify_type, notify_data):
-        self._ready.wait()
+    def _process_new_osd_map(self):
+        old_osdmap = self._latest_osdmap
+        self._latest_osdmap = self.get_osdmap()
+        assert old_osdmap
+        assert self._latest_osdmap
+
+        self.log.info("Processing OSDMap change {0}..{1}".format(
```
Comment (tchaikov) [inline src/pybind/mgr/progress/module.py:640]: ditto.

File: src/pybind/mgr/progress/module.py [hunk 11]
```diff
@@ -607,53 +620,59 @@ def _pg_state_changed(self, pg_dump):
...
+    def _process_new_pg_map(self):
+        # if there are no events we will skip this here to avoid
+        # expensive get calls
+        if len(self._events) == 0:
+            return
+
+        global_event = False
+        data = self.get("pg_stats")
+        ready = self.get("pg_ready")
+        for ev_id in list(self._events):
```
Comment (tchaikov) [inline src/pybind/mgr/progress/module.py:654]: what's the benefit of converting the keys of the dict to a list before iterating through it?

File: src/pybind/mgr/progress/module.py [hunk 12]
```diff
@@ -607,53 +620,59 @@ def _pg_state_changed(self, pg_dump):
...
+            ev = self._events[ev_id]
```
Comment (tchaikov) [inline src/pybind/mgr/progress/module.py:655]: if what we care are but the values of the dict. why not put
```py
for ev in self._events.values():
  # ...
```

**Comment by jdurgin** [comment]
[progress: improve scalability](https://trello.com/c/yMzy1mkL/600-progress-improve-scalability)

**Comment by kamoltat** [comment]
18/18 passed on teuthology so I think should be fine for fixing the global event bug
https://pulpito.ceph.com/ksirivad-2021-06-04_22:59:43-rados:mgr-wip-ksirivad-fix-bug-49988-distro-basic-smithi/

File: src/pybind/mgr/progress/module.py (hunk 8 above)
Comment (kamoltat) [reply src/pybind/mgr/progress/module.py:465]: just making the space consistent with the line above

**Comment by tchaikov** [comment]
@kamoltat i don't think all the comments were addressed. could you please leave the unresolved ones unresolved instead of marking all of them "resolved" ?

**Comment by kamoltat** [comment]
> @kamoltat i don't think all the comments were addressed. could you please leave the unresolved ones unresolved instead of marking all of them "resolved" ?

oops sorry, sure thing!

File: src/pybind/mgr/progress/module.py (hunk 4 above)
Comment (kamoltat) [reply src/pybind/mgr/progress/module.py:215]: you're right, the types of both are already int

File: src/pybind/mgr/progress/module.py (hunk 7 above)
Comment (kamoltat) [reply src/pybind/mgr/progress/module.py]: this change no longer exists

File: src/pybind/mgr/progress/module.py (hunk 2 above)
Comment (kamoltat) [reply src/pybind/mgr/progress/module.py]: change no longer exists

File: src/pybind/mgr/progress/module.py (hunk 5 above)
Comment (kamoltat) [reply src/pybind/mgr/progress/module.py]: change no longer exists

File: src/pybind/mgr/progress/module.py (hunk 11 above)
Comment (kamoltat) [reply src/pybind/mgr/progress/module.py]: there is runtime error for when
dictionary change during iteration.
Fix by just making a copy of the
dictionary using list()

File: src/pybind/mgr/progress/module.py (hunk 12 above)
Comment (kamoltat) [reply src/pybind/mgr/progress/module.py]: same reason as previous comment about runtime error when items in dictionary are changed

**Comment by kamoltat** [comment]
@tchaikov should be fine now?

**Review by jdurgin — APPROVED** [review]

File: src/pybind/mgr/progress/module.py (hunk 7 above)
Comment (tchaikov) [reply src/pybind/mgr/progress/module.py]: although it is unrelated to this PR, i'd keep this comment unresolved. as "5" is the interval between each call to persist the events, from user's perspective, the higher "frequency" the more frequently we persist the events. so this description is not user friendly.

**Comment by tchaikov** [comment]
> @tchaikov should be fine now?

i haven't reviewed your change thoroughly . so i am not able to claim it's "fine". just wanted to improve its quality by nit picking...

the last thing, the commit message reads

> Fixes: Fixes: https://tracker.ceph.com/issues/49988

probably you could one of the "Fixes: "?
